test/models/order_test.rb

```diff
@@ -0,0 +1,177 @@
+      order.errors.messages.must_include :city
+    end
+
+    it "requires a state" do
+      order.state = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :state
+    end
+    it "requires a zipcode" do
+      order.zip_code = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :zip_code
+    end
+
+    it "requires a name on card for billing" do
+      order.name_on_card = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :name_on_card
+    end
+
+    it "requires a credit card number" do
+      order.card_number = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :card_number
+    end
+
+    it "requires expiration month" do
+      order.exp_month = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :exp_month
+    end
+
+    it "requires expiration year" do
+      order.exp_year = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :exp_year
+    end
+
+    it "requires cvv" do
+      order.cvv = nil
+      order.valid?.must_equal false
+      order.errors.messages.must_include :cvv
+    end
+
+    it "allows only predefined orderitem_status" do
+      valid_status = ['pending', 'paid']
+      valid_status.each do |status|
+        order.order_state = status
+        order.valid?.must_equal true
+      end
+    end
+
+    it "rejects invalid order_state" do
+      invalid_states = ['lost', 'shipped', 'backordered', 'banana']
+      invalid_states.each do |invalid_state|
+        order_no_state.order_state = invalid_state
+        order_no_state.valid?.must_equal false
+        order_no_state.errors.messages.must_include :order_state
+      end
+    end
+  end
+
+  describe "custom methods" do
+    describe "in_cart?" do
+      it "allow NIL values for all other fields, given order_state is pending/in-cart" do
+        cart_order = Order.new(order_state: "pending")
+        cart_order.valid?.must_equal true
+      end
+
+      it "won't allow NIL values for any fields, given order_state is not pending" do
+        cart_order = Order.new(order_state: "paid")
+        cart_order.valid?.must_equal false
+      end
+    end
+
+
+    describe "add_to_cart" do
+        before do
+          @cart = Order.create(order_state: "pending")
+        end
+
+      it "creates a new orderitem instance if the product doesn't exist in the cart yet" do
+        before = @cart.orderitems.count
+        @cart.add_to_cart({ :orderitem => { product_id: products(:five).id, quantity: 1 } })
+        @cart.orderitems.count.must_be :>, before
+      end
+
+      it "won't create a new orderitem instance if product already exists in the cart" do
+        @cart.add_to_cart({ :orderitem => { product_id: products(:five).id, quantity: 1 } })
+        before = @cart.orderitems.count
+        @cart.add_to_cart({ :orderitem => { product_id: products(:five).id, quantity: 1 } })
```

Comment:
These tests seems to be using this hash throughout so it would be good to DRY this up and use a shared variable